mitiq/calibration/calibration.py

```diff
@@ -0,0 +1,204 @@
+
+        noisy = num_circuits * num_methods * num_options
+        ideal = noisy if self.ideal_executor else 0
+        return {
+            "noisy_executions": noisy,
+            "ideal_executions": ideal,
+        }
+
+    def run_circuits(self) -> list[dict[str, Any]]:
+        """Run the calibration circuits and store the ideal and noisy
+        expectation values for each circuit in `self.results`."""
+        expvals = []
+        for circuit_data in self.circuits:
+            circuit = circuit_data.circuit
+            circuit.append(cirq.measure(circuit.all_qubits()))
+
+            noisy_results = cast(
+                MeasurementResult, self.executor._run([circuit])[0]
+            ).result
+            noisy_distribution = bitstrings_to_distribution(noisy_results)
+            ideal_results = (
+                cast(
+                    MeasurementResult, self.ideal_executor._run([circuit])[0]
+                ).result
+                if self.ideal_executor
+                else None
+            )
+            ideal_distribution = (
+                bitstrings_to_distribution(ideal_results)
+                if ideal_results
+                else {"1" * len(circuit.all_qubits()): 1.0}
+            )
+
+            mitigated: dict[str, dict[str, Any]] = {
+                technique: {"results": [], "method_improvement_factor": None}
+                for technique in self.settings.techniques
+            }
+            for strategy in self.settings.make_strategies():
+                most_probable_bitstring = max(
+                    ideal_distribution, key=ideal_distribution.get
+                )
+                expval_executor = bitstring_executor_to_expval_executor(
+                    self.executor, most_probable_bitstring
+                )
+                mitigated_expval = strategy.mitigation_function(
+                    circuit, expval_executor
+                )
+                mitigated[strategy.technique]["results"].append(
+                    {
+                        "circuit_type": circuit_data.type,
+                        "mitigated_value": mitigated_expval,
+                        **strategy.technique_params,
+                    }
+                )
+            expvals.append(
+                {
+                    "noisy_dist": noisy_distribution,
+                    "ideal_dist": ideal_distribution,
+                    "mitigated": mitigated,
+                }
+            )
+        return expvals
+
+    def compute_improvements(
+        self, experiment_results: list[dict[str, Any]]
+    ) -> None:
+        """Compute the improvement factors for each calibration circuit that
+        was run."""
+        for result in experiment_results:
+            ideal_dist = result["ideal_dist"]
+            ideal_expval = max(ideal_dist.values())
+            noisy_dist = result["noisy_dist"]
+            noisy_expval = max(noisy_dist.values())
```

Comment:
```suggestion
            ideal_value = result["ideal_value"]
            noisy_value = result["noisy_value"]
```

Instead of taking the maximum of the observed frequencies we should take the frequency of `bitstring_to_measure`.  This is not necessarily the same thing, especially if noise is very strong.

Reply:
Good catch, I like this refactor.